File: part2/Akka Cluster/src/main/scala/GUI/SelectionManager.scala
package GUI

import model.PuzzleActor

import java.awt.Color
import javax.swing.{BorderFactory, JButton}

object SelectionManager:
    @FunctionalInterface private[GUI] trait Listener {
        def onSwapPerformed(): Unit
    }

class SelectionManager(val actor: PuzzleActor):

    var selectionActive: Boolean = false
    var selectedTile: Tile = null
    var tileToBeSelected: Tile = null
    
    def selectTile(tile: Tile, listener: SelectionManager.Listener, toSend: Boolean): Unit =
        if (toSend)
            actor.sendSelection(tile.currentPosition)
        if (selectionActive)
            selectionActive = false
            swap(selectedTile, tile)
            listener.onSwapPerformed()
        else
            selectionActive = true
            selectedTile = tile
            val button: JButton = actor.puzzle.board.getComponent(tile.currentPosition).asInstanceOf[JButton]
            button.setBorder(BorderFactory.createLineBorder(Color.red))

    private def swap(t1: Tile, t2: Tile): Unit =
        val pos = t1.getCurrentPosition
        t1.setCurrentPosition(t2.getCurrentPosition)
        t2.setCurrentPosition(pos)
        

File: part2/Java RMI/src/main/java/org/example/puzzle/SelectionManager.java
package org.example.puzzle;

import java.io.Serializable;
import java.util.List;

public class SelectionManager implements Serializable {

	private Tile selectedTile;

	public SelectionManager() {

	}

	public void selectTile(List<Tile> tiles, final Tile tile, final Listener listener) {
		Tile tl = tiles
				.stream()
				.filter(t -> t.getOriginalPosition() == tile.getOriginalPosition())
				.findAny()
				.get();
		if (selectedTile != null) {
			tl.setSelected(false);
			selectedTile.setSelected(false);
			swap(selectedTile, tl);
			listener.onSwapPerformed();
			selectedTile = null;
		} else {
			tl.setSelected(true);
			selectedTile = tl;
		}
	}

	private void swap(final Tile t1, final Tile t2) {
		int pos = t1.getCurrentPosition();
		t1.setCurrentPosition(t2.getCurrentPosition());
		t2.setCurrentPosition(pos);
	}
	
	@FunctionalInterface
	public interface Listener{
		void onSwapPerformed();
	}
}
